src/Microsoft.VisualStudio.LanguageServices.Razor/Editor/RazorDirectiveCompletionProvider.cs

```diff
@@ -0,0 +1,185 @@
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Projection;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor.Editor
+{
+    [System.Composition.Shared]
+    [Export(typeof(CompletionProvider))]
+    [ExportMetadata("Language", LanguageNames.CSharp)]
+    internal class RazorDirectiveCompletionProvider : CompletionProvider
+    {
+        // Internal for testing
+        internal static readonly string DescriptionKey = "Razor.Description";
+
+        private static readonly IEnumerable<DirectiveDescriptor> DefaultDirectives = new[]
+        {
+            CSharpCodeParser.AddTagHelperDirectiveDescriptor,
+            CSharpCodeParser.RemoveTagHelperDirectiveDescriptor,
+            CSharpCodeParser.TagHelperPrefixDirectiveDescriptor,
+        };
+        private readonly RazorCodeDocumentProvider _codeDocumentProvider;
+
+        [ImportingConstructor]
+        public RazorDirectiveCompletionProvider(RazorCodeDocumentProvider codeDocumentProvider)
+        {
+            if (codeDocumentProvider == null)
+            {
+                throw new ArgumentNullException(nameof(codeDocumentProvider));
+            }
+
+            _codeDocumentProvider = codeDocumentProvider;
+        }
+
+        public override Task<CompletionDescription> GetDescriptionAsync(Document document, CompletionItem item, CancellationToken cancellationToken)
+        {
+            if (document == null)
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            if (item == null)
+            {
+                throw new ArgumentNullException(nameof(item));
+            }
+
+            var descriptionContent = new List<TaggedText>();
+            if (item.Properties.TryGetValue(DescriptionKey, out var directiveDescription))
+            {
+                var descriptionText = new TaggedText(TextTags.Text, directiveDescription);
+                descriptionContent.Add(descriptionText);
+            }
+
+            var completionDescription = CompletionDescription.Create(descriptionContent.ToImmutableArray());
+            return Task.FromResult(completionDescription);
+        }
+
+        public override Task ProvideCompletionsAsync(CompletionContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (!context.Document.FilePath.EndsWith(".cshtml", StringComparison.OrdinalIgnoreCase))
+            {
+                // Not a Razor file.
+                return Task.FromResult(0);
+            }
+
+            if (!_codeDocumentProvider.TryGetFromDocument(context.Document, out var codeDocument))
+            {
+                // A Razor code document has not yet been associated with the document.
+                return Task.FromResult(0);
+            }
+
+            var syntaxTree = codeDocument.GetSyntaxTree();
```

Comment:
should we null check this?

Reply:
Sure to be extra extra sure. However, given our current scenarios it will never be `null`.